### ISSUE TITLE
Performance issues after repeated `push_notebook` calls

### ISSUE THREAD

**Issue by stanleyguan** [bokeh/bokeh#4247]
### Description of expected behavior and the observed behavior

Expected: Reasonably performant visualizations when pushing many simple updates to a single plot in a Jupyter Notebook. In addition, time taken to push updates to a new plot shouldn't be affected by the number of pushes to an older plot.

Observed: Performance appears to degrade linearly with the number of previous `push_notebook` calls.
### Self-contained example code that generates the issue (if possible)

```
import bokeh
import bokeh.plotting as bk
bokeh.io.output_notebook()
fig = bk.figure()
l = fig.line([], [])
bk.show(fig)
```

```
%%time
for i in range(1000):
    l.data_source.stream({'x':[i],'y':[i]})
    bokeh.io.push_notebook()
```

Run the above loop a couple of times in its own cell and you should notice performance degrade from about 100 points a second to 50 points a second after a few loops. The degraded performance does not restore by pushing to a new, different plot. Performing a kernel reset restores the starting performance.
### Stack traceback and/or browser developer tools console output

Nothing unusual spotted.
### Version Information

```
Python version      :  2.7.11 |Anaconda 2.5.0 (x86_64)| (default, Dec  6 2015, 18:57:58) 
IPython version     :  4.1.2
Bokeh version       :  0.11.1
BokehJS static path :  ~/lib/python2.7/site-packages/bokeh/server/static
```

**Comment by bryevdv** [comment]
I think this is because you are appending an unbounded number of points, without specifying any rollover. There is a cost associated with rendering that is proportional to the number of points. If I try specifying a rollover of 100:
```
%%time
for i in range(1000):
    l.data_source.stream({'x':[i],'y':[i]}, 100)
    bokeh.io.push_notebook()
```
And run that code over and over, the reported time stays the same (about 1.9 seconds on my machine).

**Comment by bryevdv** [comment]
By way of very rough comparison, if I run `examples/plotting/file/color_scatter.py` , with `BOKEH_LOG_LEVEL=trace` I can see:

<img width="646" alt="screen shot 2016-04-29 at 3 55 39 pm" src="https://cloud.githubusercontent.com/assets/1078448/14929425/f0a7660c-0e22-11e6-9ef8-3e9b9ebea2a6.png">

This is not a straight comparison, as this example is circles instead of lines, and it has about 4k points, but hopefully it illustrates that rendering is not free.

**Comment by stanleyguan** [comment]
Using the `rollover` argument does seem to stop performance from degrading - when running the second cell repeatedly. However, when I rerun the first cell and then run the second again, the performance degrades again. Is that expected?

Also I couldn't find documentation for the `rollover` argument.

**Comment by bryevdv** [comment]
Expected, probably not. It will require some further investigation. I'm not sure when I will be able to dive into it, I have several weeks of talks and travel upcoming. If you would like to try to collect more information with a little "printf debugging" in the mean time, that could be very helpful. The relevant code is here:

https://github.com/bokeh/bokeh/blob/master/bokeh/core/property_containers.py#L163-L184

The Bokeh docs are fairly extensive, but the project is also large and the team small, so there are certainly still documentation gaps. The streaming API was a very late addition to the `0.11.1` release and it was decided that making the feature available with only supporting examples for documentation was more valuable than omitting it completely. If you have time, small PR to add a docstring according to our usual docstring style would make sure information at lest gets automatically added to the reference guide on future releases, and certainly be appreciated. The method is here:

https://github.com/bokeh/bokeh/blob/master/bokeh/models/sources.py#L225

**Comment by bryevdv** [comment]
OK the current status with `0.12` is the executing both cells repeatedly results in about a ~1 second increase in time every exectution. My guess is that lot state from old comms handles is being kept around and synced. Propose investigating cell hooks or similar to clear our comms handles for cell outputs that get overwritten.

**Comment by davidbrochart** [comment]
I can still see the issue in `0.12.4`. Any progress?

**Comment by bryevdv** [comment]
Progress has been made at  easily embedding Bokeh server apps in notebooks. In the latest `0.12.5` dev build you can do this:

https://github.com/bokeh/bokeh/blob/master/examples/howto/server_embed/notebook_embed.ipynb

It's possible that improvements can still be made to `push_notebook` however *in principle* it will never be as robust or performant as the technique in notebook above, which uses the Bokeh web socket protocol directly. So my recommendation is to use that approach instead in any case.

**Comment by davidbrochart** [comment]
Thanks, I'll look at it.

**Comment by bryevdv** [comment]
This should be fixed by #6945 The following version:
```
%%time
for i in range(500):
    l.data_source.stream({'x':[i],'y':[i]}, rollover=200)
```
runs in ~550ms even after dozens of executions
